RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,206 @@
+
+    """This class stores information about one movement,
+    i.e. a car moving from position *from_index* to position *to_index*.
+    """
+
+    def __init__(self, car, from_index, to_index):
+        self.car = car
+        self.from_index = from_index
+        self.to_index = to_index
+
+    def __str__(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        return move_str
+
+class RearrangeCars(object):
+
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Stores and updates min_num_steps_left that shows min number of moves left
+        in order to avoid comparison current and final order lists that takes O(n)
+    '''
+
+    def __init__(self, init_car_order, final_car_order):
+
+        '''Initializes class instance with initial, final order and
+        stores the position of empty parking space in the current car order
+        Memory is O(n) to store 3 order lists, 1 car_position_dict and a set of
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
+        in O(1) unarranged car
+
+        Arguments:
+           init_car_order: a valid non-empty list of cars + 1 \
+           empty_place given in their parking place order
+
+           final_car_order: a valid non-empty list of cars + 1 \
+           empty_place according to the desired order
+
+        Returns:
+            None
+        '''
+
+        self.initial_car_order = init_car_order
+        self.final_car_order = final_car_order
+        self.print_console = True
+        self._variables_to_init_state()
+
+    def print_moves_log_file(self, moves_list):
+
+        '''Prints a list of moves to the log file
+        
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+        
+        info_message = 'Number moves is {0} > 1000, logging them..'.format(len(moves_list))
+        print info_message
+        logging.info('Initial car order:')
+        logging.info(self.initial_car_order)
+        logging.info('Final car order:')
+        logging.info(self.final_car_order)
+        logging.info('Moves:')
+        for move in moves_list:
+            logging.info(move.__str__())   
+     
+    def print_moves_console(self, moves_list):
+        
+        '''Prints a list of moves
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+        
+        print 'Initial car order:', self.initial_car_order
+        print 'Final car order:', self.final_car_order

```

Comment:
We could use some neat formatting (see f strings), but that applies only to Python 3 code